services/archiver/archiver.go

```diff
@@ -0,0 +1,309 @@
+		Prefix: setting.Repository.PrefixArchiveFiles,
+	}); err != nil {
+		log.Error("Download -> CreateArchive "+tmpArchive.Name(), err)
+		return
+	}
+
+	// Now we copy it into place
+	if destArchive, err = os.Create(r.archivePath); err != nil {
+		log.Error("Unable to open archive " + r.archivePath)
+		return
+	}
+	_, err = io.Copy(destArchive, tmpArchive)
+	destArchive.Close()
+	if err != nil {
+		log.Error("Unable to write archive " + r.archivePath)
+		return
+	}
+
+	// Block any attempt to finalize creating a new request if we're marking
+	r.archiveComplete = true
+}
+
+// ArchiveRepository satisfies the ArchiveRequest being passed in.  Processing
+// will occur in a separate goroutine, as this phase may take a while to
+// complete.  If the archive already exists, ArchiveRepository will not do
+// anything.  In all cases, the caller should be examining the *ArchiveRequest
+// being returned for completion, as it may be different than the one they passed
+// in.
+func ArchiveRepository(request *ArchiveRequest) *ArchiveRequest {
+	// We'll return the request that's already been enqueued if it has been
+	// enqueued, or we'll immediately enqueue it if it has not been enqueued
+	// and it is not marked complete.
+	archiveMutex.Lock()
+	if rExisting := getArchiveRequest(request.repo, request.commit, request.archiveType); rExisting != nil {
+		archiveMutex.Unlock()
+		return rExisting
+	}
+	if request.archiveComplete {
+		archiveMutex.Unlock()
+		return request
+	}
+
+	archiveInProgress = append(archiveInProgress, request)
+	archiveMutex.Unlock()
+	go func() {
+		// Wait to start, if we have the Cond for it.  This is currently only
+		// useful for testing, so that the start and release of queued entries
+		// can be controlled to examine the queue.
+		if archiveQueueStartCond != nil {
+			archiveQueueMutex.Lock()
+			archiveQueueStartCond.Wait()
+			archiveQueueMutex.Unlock()
+		}
+
+		// Drop the mutex while we process the request.  This may take a long
+		// time, and it's not necessary now that we've added the reequest to
+		// archiveInProgress.
+		doArchive(request)
+
+		if archiveQueueReleaseCond != nil {
+			archiveQueueMutex.Lock()
+			archiveQueueReleaseCond.Wait()
+			archiveQueueMutex.Unlock()
+		}
+
+		// Purge this request from the list.  To do so, we'll just take the
+		// index at which we ended up at and swap the final element into that
+		// position, then chop off the now-redundant final element.  The slice
+		// may have change in between these two segments and we may have moved,
+		// so we search for it here.  We could perhaps avoid this search
+		// entirely if len(archiveInProgress) == 1, but we should verify
+		// correctness.
+		archiveMutex.Lock()
+		defer archiveMutex.Unlock()
+		// Wake up all other goroutines that may be waiting on a request to
+		// complete.  They should all wake up, see if that particular request
+		// is complete, then return to waiting if it is not.
+		archiveCond.Broadcast()
```

Comment:
This is what channels (`chan`) are for 😉

Reply:
... but now that I read this particular comment, the solution makes more sense and doesn't involve Cond at all. :-)